Documents/AnalyticsToolDesignDocument.md

```diff
@@ -0,0 +1,108 @@
+- The **Git index** is git's internal list of files which it is tracking for creating  the next `commit`. It is useful to this project by means of comparing it to the list of **placeholders** to deduce if those files should truly be hydrated or if it may have occurred incidentally.
+- The **file system** must be carefully considered in the case of working with a mounted repository. VFS4G intercepts file I/O calls at the kernel level to inject its own behavior (the projected file system and if needed a network call to pull down the file which the system asked for in the correct format). Because of this, enumerating directories can cause additional data to be pulled down from the remote by hydrating extra files or further hydrating those that already exist on disk. Accordingly, care must be taken when interacting with the file system lest the tool designed to help users increase enlistment heath end up causing more problems.
+
+## Components
+
+- **API** - The server component of the client-server relationship. Responsible for the logic and computation of statistical data, as well as the assembly and sending of a **json statistics object** over the named pipe to the client. It sits within the `InProgressMount` process and interacts with the VFS4G internals
+- **Verb** - The `gvfs statistics` verb will be the component that handles the client side interaction with the API. When run, it will collect and store the statistics from the enlistment to either be displayed immediately or collected for trend purposes to then provide understanding in the future. This is also the point where data can be sent to telemetry for our internal analysis. The verb will also output the data to the console with a tool to help visualize and understand the data. It will output the analytics data to the command line and will also be used during the design period to test the API and verb
+- **Utility tools for dehydration** - *Stretch goal* | After the analytics are collected, these components will provide utilities to help the user get their enlistment back to a healthier state. The ideal goal is to have VFS4G suggest uses to the user (EG: `This directory was hydrated by Sublime text editor. You can dehydrate it by running "gvfs dehydrate --confirm -d /path/to/repo"`)
+- **Additional visualization tools** - *Stretch goal* | With a comprehensive API in place, the possibility opens up for different visualization tools to be created that consume the json data and display it to the user in different ways, potentially even eventually interfacing with the utility tools for an interactive GUI
+
+## Example of CLI Tool Usage
+
+```terminal
+$ gvfs statistics
+
+Repository statistics
+Total paths tracked by git:        3,123,456 |  100%
+Total number of placeholders:        123,456 |    4%
+Total number of modififed paths:       1,234 |   <1%
+
+Total hydration percentage:                       4%
+Most hydrated directories:

```

Comment:
Oh, this summary of the few most-hydrated directories is so much better than the full list of directories I had in my head. Awesome.

Make a note for later: when do we stop this list? Perhaps include directories whose hydration percentage is at least the total hydration percentage? That way, we can easily see if they have 25% of their root directories 100% hydrated, or have 100% of their root directories 25% hydrated.

Reply:
Thanks a good idea! I hadn't thought of that, I'll note that down in the document as the default behavior. This morning I'm working on adding the other features and parameters that me and Jameson briefly discusses and I think will be cool. One of them would be a `--hydration` parameter which defaults to the root and just shows more information about the hydration of different directories, or could be pointed at a cone of the tree and then would go more in depth from there.